### DISCUSSION TITLE
How to support older browsers?

### DISCUSSION THREAD

**Discussion by osdiab** [chakra-ui/panda#844] [Q&A] [answered]
### Subject

Reason not to use this

### Description

Under the reasons not to use Panda, nothing is stated about needing to support older browsers (by older I mean browsers released before 2022, since layers themselves aren't supported there, including Safari on older iOS/Mac versions, which still has a sizable userbase). It would be good to have some documentation on what browsers are actually supported by Panda.

**Comment by segunadebayo** [comment, marked as answer]
Hey @osdiab,

There's a PostCSS Plugin you can use to support older browsers.
https://www.npmjs.com/package/@csstools/postcss-cascade-layers

We'll update the docs as needed

**Reply by IanVS** [reply]
Have the docs been updated?  The postCSS plugin only works if all of the code is sent to postcss at the same pass (https://github.com/csstools/postcss-plugins/issues/376#issuecomment-1126511448), so I guess this would work so long as panda creates one big file.  But how would that work with code splitting?  I'm just learning about Panda.css, so apologies if this is off-base.

**Reply by itsjavi** [reply]
@segunadebayo can you pls provide an example of how to set it up using NextJS 13 (app router)?
I tried this and the generated `layout.css` (I import global.css in layout.tsx) still has `@layer` styles (not supported in browsers from some years ago like Chrome 91).

`browserlist` is also ignored. I don't know if it's a problem of NextJS or Panda CSS.

```js
/** @type {import('postcss-load-config').Config} */
const config = {
  plugins: [
    '@pandacss/dev/postcss',
    '@csstools/postcss-cascade-layers'
  ],
}

module.exports = config
```

also tried this without luck:
```js
/** @type {import('postcss-load-config').Config} */
const config = {
  plugins: [
    '@pandacss/dev/postcss',
    [
      'postcss-preset-env',
      {
        stage: 2,
        features: {
          'color-mix': true,
          'cascade-layers': true,
        },
      },
    ],
  ],
}

module.exports = config
```

**Reply by huynhhuutai** [reply]
@segunadebayo Could you share an example for the config?
Same as @itsjavi I have tried many approaches but not been able to address this
@itsjavi Have you figured it out?

**Reply by astahmer** [reply]
since the opening of this issue, we shipped a `polyfill` config option that enables it
https://panda-css.com/docs/references/config#polyfill

**Comment by wmertens** [comment]
So this plugin converts all layers to more complex old-style rules, correct?

Ideally, only old browsers get these rules, no? Or is the overhead minimal?

Suppose you have the regular `app.css` and the processed `app-nolayers.css`, then could the CSS be done as follows?

- load `app-nolayers.css` inside a `<noscript>` tag
- inside a `<script>` tag, check if the css supports layers, and load `app.css` if so, otherwise load `app-nolayers.css`.

Is this a good idea or overkill?

For Qwik, CSS gets inlined automatically, so this behavior will need to be implemented at the SSR level, or else it could load the nolayer css on top of the layer css in case of noscript. I imagine that would be problematic?

**Comment by itsjavi** [comment]
Can we have a configuration option to opt-out CSS layers 🙏, or would it need too much refactoring?

It's giving me some headaches to make it work properly between many projects, specially when using many projects that use Panda together, and also as a UI library maintainer.

**Reply by wmertens** [reply]
@itsjavi looking at the code I think that's unlikely. Instead, use the postcss plugin to convert to no-layers syntax.
